src/pyclass/guard.rs

```diff
@@ -0,0 +1,745 @@
+    }
+
+    /// Gets a `PyClassGuard<T::BaseType>`.
+    ///
+    /// With the help of this method, you can get hold of instances of the
+    /// super-superclass when needed.
+    ///
+    /// # Examples
+    /// ```
+    /// # use pyo3::prelude::*;
+    /// #[pyclass(subclass)]
+    /// struct Base1 {
+    ///     name1: &'static str,
+    /// }
+    ///
+    /// #[pyclass(extends=Base1, subclass)]
+    /// struct Base2 {
+    ///     name2: &'static str,
+    /// }
+    ///
+    /// #[pyclass(extends=Base2)]
+    /// struct Sub {
+    ///     name3: &'static str,
+    /// }
+    ///
+    /// #[pymethods]
+    /// impl Sub {
+    ///     #[new]
+    ///     fn new() -> PyClassInitializer<Self> {
+    ///         PyClassInitializer::from(Base1 { name1: "base1" })
+    ///             .add_subclass(Base2 { name2: "base2" })
+    ///             .add_subclass(Self { name3: "sub" })
+    ///     }
+    ///     fn name(slf: PyClassGuard<'_, Self>) -> String {
+    ///         let subname = slf.name3;
+    ///         let super_ = slf.into_super();
+    ///         format!("{} {} {}", super_.as_super().name1, super_.name2, subname)
+    ///     }
+    /// }
+    /// # Python::attach(|py| {
+    /// #     let sub = Py::new(py, Sub::new()).unwrap();
+    /// #     pyo3::py_run!(py, sub, "assert sub.name() == 'base1 base2 sub'")
+    /// # });
+    /// ```
+    pub fn into_super(self) -> PyClassGuard<'a, U> {
+        PyClassGuard {
+            ptr: std::mem::ManuallyDrop::new(self).ptr,
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<T: PyClass> Deref for PyClassGuard<'_, T> {
+    type Target = T;
+
+    #[inline]
+    fn deref(&self) -> &T {
+        // SAFETY: `PyClassObject<T>` constains a valid `T`, by construction no
+        // mutable alias is enforced
+        unsafe { &*self.as_class_object().get_ptr().cast_const() }
+    }
+}
+
+impl<'a, 'py, T: PyClass> FromPyObjectBound<'a, 'py> for PyClassGuard<'a, T> {
+    fn from_py_object_bound(obj: Borrowed<'a, 'py, crate::PyAny>) -> crate::PyResult<Self> {
+        Self::try_from_class_object(obj.downcast()?.get_class_object()).map_err(Into::into)
+    }
+}
+
+impl<'a, 'py, T: PyClass> IntoPyObject<'py> for PyClassGuard<'a, T> {
+    type Target = T;
+    type Output = Borrowed<'a, 'py, T>;
+    type Error = Infallible;
+
+    #[inline]
+    fn into_pyobject(self, py: crate::Python<'py>) -> Result<Self::Output, Self::Error> {
+        (&self).into_pyobject(py)
+    }
+}
+
+impl<'a, 'py, T: PyClass> IntoPyObject<'py> for &PyClassGuard<'a, T> {
+    type Target = T;
+    type Output = Borrowed<'a, 'py, T>;
+    type Error = Infallible;
+
+    #[inline]
+    fn into_pyobject(self, py: crate::Python<'py>) -> Result<Self::Output, Self::Error> {
+        // SAFETY: `ptr` is guaranteed to be valid for 'a and points to an
+        // object of type T
+        unsafe { Ok(Borrowed::from_ptr_unchecked(py, self.ptr.as_ptr()).downcast_unchecked()) }

```

Comment:
Maybe a `Borrowed::from_non_null` internal helper will make this better to read?


```suggestion
        unsafe { Ok(Borrowed::from_non_null(py, self.ptr).downcast_unchecked()) }
```